Using only the docsfile to build the text index

Flixtastic opened the pull request:
This PR is to enable only using the docsfile during text index building without the need of a complicated often misleading wordsfile. The wordsfile still can be used to add entities so ql:contains-entity still works even when using the docsfile to build the word part of the text index.

Flixtastic committed:
Splitting the parsing of the wordsfile and the literals so one can have more control over it.

Flixtastic committed:
Changed the text index building to take the wordsfile and docsfile separately and added the option to build the text index only by using the docsfile. Changed the internal parsing for text index building to parse the docsfile if wished for. The behaviour isn't tested yet.

Flixtastic committed:
Reworked Info message.

Flixtastic committed:
Added simple testing for docsfile text index building. Changed the saving of nofNonLiterals of text index to the saving of the last text record index of non literals.

Flixtastic committed:
Implemented using the wordsfile only to add entities when building the text index with the docsfile. This is not being tested yet but the methods to test were changed to make it possible.

Flixtastic committed:
Added Tests for building text index using the docsfile and also adding entities with the wordsfile.

Flixtastic committed:
Merge remote-tracking branch 'upstream/master' into docsfile-only-text-index

# Conflicts:
#	src/index/IndexImpl.Text.cpp

Flixtastic committed:
Merge remote-tracking branch 'upstream/master' into docsfile-only-text-index

# Conflicts:
#	src/index/Index.cpp
#	src/index/Index.h
#	src/index/IndexBuilderMain.cpp
#	src/index/IndexImpl.Text.cpp
#	src/index/IndexImpl.h
#	test/util/IndexTestHelpers.cpp

Flixtastic committed:
Small Sonar changes

joka921 commented on src/index/IndexBuilderMain.cpp:
```diff
@@ -321,22 +330,31 @@ int main(int argc, char** argv) {
+          docsfile.empty() ? std::nullopt
+                           : std::optional<const string>(docsfile),
+          addWordsFromLiterals, useWordsFromDocsfile, addEntitiesFromWordsfile,
+          getTextScoringMetricFromString(scoringMetric),
```
We'll revisit the exact wordings here at the end.

joka921 commented on src/index/IndexImpl.cpp:
```diff
@@ -1144,7 +1144,8 @@ void IndexImpl::readConfiguration() {
   loadDataMember("num-objects", numObjects_, NumNormalAndInternal{});
   loadDataMember("num-triples", numTriples_, NumNormalAndInternal{});
-  loadDataMember("num-non-literals-text-index", nofNonLiteralsInTextIndex_, 0);
+  loadDataMember("last-text-record-index-of-non-literals",
```
Is it bad if this member is not found and thus 0? 
Because it will be, if we load an old index.

Flixtastic replied:
If checked correctly it is only relevant for testing since it doesn't occur anywhere else directly.

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -116,26 +99,30 @@ void IndexImpl::logEntityNotFound(const string& word,
+                           (wordsFile.has_value() || useDocsFileForVocabulary);
+  AD_CORRECTNESS_CHECK(addWordsFromFiles || addWordsFromLiterals);
+  AD_CORRECTNESS_CHECK(!addEntitiesFromWordsFile || useDocsFileForVocabulary);
+  string wordsFileString = wordsFile.value_or("");
+  string docsFileString = docsFile.value_or("");
```
If you do this so early, then the ` wordsFileString` should be called ` wordsFile`  and the optional should be called something else.

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -246,29 +237,52 @@ void IndexImpl::addTextFromOnDiskIndex() {
     }
+  };
+  if (useDocsFileForVocabulary) {
+    const auto& localeManager = textVocab_.getLocaleManager();
+    for (const auto& line : DocsFileParser{file, localeManager}) {
+      for (const auto& word :
+           tokenizeAndNormalizeText(line.docContent_, localeManager)) {
+        auto lineToProcess = WordsFileLine{
```
Can you make this a single view in a separate function s.t. it becomes
`for (const auto& line : getWordsLineFormDocsfile(file, localManager) { processLine(line)}`

That function can initially be a generator, but then we refactor it using `ql::views::transform`  and ` ql::views::join`  etc.

Flixtastic replied:
ee17c033cd1f6060905f988d310f4bd79a227cc8

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -246,29 +237,52 @@ void IndexImpl::addTextFromOnDiskIndex() {
+                                             bool useDocsFileForVocabulary,
+                                             bool addEntitiesFromWordsFile,
                                              TextVec& vec) {
+  AD_CONTRACT_CHECK(!addEntitiesFromWordsFile || useDocsFileForVocabulary);
+  AD_CONTRACT_CHECK(!(wordsFile.empty() && addEntitiesFromWordsFile));
+  AD_CONTRACT_CHECK(!(docsFile.empty() && useDocsFileForVocabulary));
```
These checks (also the ones above) can all be encapsulated in the ` TextIndexConfig`  struct.

Flixtastic replied:
09036fd1f6bc7b167e47045b24217ef626998542

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -298,6 +312,76 @@ void IndexImpl::processWordsForInvertedLists(const string& contextFile,
+      auto docsFileIterator = docsFileParser.begin();
+      auto wordsFileParser = WordsFileParser{wordsFile, localeManager};
+      auto wordsFileIterator = wordsFileParser.begin();
+      // Get the respective lines
+      WordsFileLine currentWordsFileLine = *wordsFileIterator;
+      DocsFileLine currentDocsFileLine = *docsFileIterator;
+
+      // Iterate over docsfile and wordsfile in parallel. The wordsfile IDs are
+      // in smaller increments than the docsfile IDs in general and the
+      // wordsfile ID belongs to the next largest or equal docsfile ID. Because
+      // of this the wordsfile is advanced until the wordsfile ID is larger than
+      // the docsfile ID. During this the entities from the wordsfile are added
+      // to the respective context in this case the respective document. Once
+      // the wordsfile ID is larger than the docsfile ID the docsfile line is
+      // processed and so on.
+      while (docsFileIterator != docsFileParser.end()) {
+        currentDocsFileLine = *docsFileIterator;
+        while (wordsFileIterator != wordsFileParser.end()) {
+          currentWordsFileLine = *wordsFileIterator;
+          if (currentDocsFileLine.docId_.get() <
+              currentWordsFileLine.contextId_.get()) {
+            break;
+          }
+          if (currentWordsFileLine.isEntity_) {
+            WordsFileLine lineToProcess = currentWordsFileLine;
+            lineToProcess.contextId_ =
+                TextRecordIndex::make(currentDocsFileLine.docId_.get());
+            processLine(lineToProcess);
+          }
+          ++wordsFileIterator;
+        }
+        for (const auto& word : tokenizeAndNormalizeText(
+                 currentDocsFileLine.docContent_, localeManager)) {
+          WordsFileLine lineToProcess = {
+              word, false,
+              TextRecordIndex::make(currentDocsFileLine.docId_.get()), 0,
+              false};
+          processLine(lineToProcess);
+        }
+        ++docsFileIterator;
```
That function is too large and too complex. 
Especially the first case (which is complex) should have its own place.
Considder writing functions that receive the ` processLine`  as a callback.

Flixtastic replied:
175d16d6e0f172c675fdfb11d6bbedc284380fde

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -298,6 +312,76 @@ void IndexImpl::processWordsForInvertedLists(const string& contextFile,
+        ++docsFileIterator;
+      }
+    } else {
+      for (const auto& line : DocsFileParser{docsFile, localeManager}) {
+        for (const auto& word :
+             tokenizeAndNormalizeText(line.docContent_, localeManager)) {
+          WordsFileLine lineToProcess = {
+              word, false, TextRecordIndex::make(line.docId_.get()), 0, false};
```
This is code duplication to above where I suggested abstracting this out.

Flixtastic replied:
ee17c033cd1f6060905f988d310f4bd79a227cc8

joka921 reviewed (changes requested):
An initial round of reviews.

Make sure to merge in the current master, where I have implemented simpler index testings.

Flixtastic committed:
Merge remote-tracking branch 'upstream/master' into docsfile-only-text-index

# Conflicts:
#	test/ExecuteUpdateTest.cpp
#	test/engine/SpatialJoinAlgorithmsTest.cpp
#	test/engine/SpatialJoinTestHelpers.h
#	test/engine/TextIndexScanForWordTest.cpp
#	test/util/IndexTestHelpers.cpp
#	test/util/IndexTestHelpers.h

Problems in TextIndexScanForEntityTest.cpp that were no conflicts.

Flixtastic committed:
Implement a TextScoringConfig

Flixtastic commented on src/index/Index.h:
```diff
@@ -225,7 +226,7 @@ class Index {
   size_t getNofWordPostings() const;
   size_t getNofEntityPostings() const;
-  size_t getNofNonLiteralsInTextIndex() const;
+  size_t getLastTextRecordIndexOfNonLiterals() const;
```
This is important for the weird retrieval of literals during text index testing.
When using literals as well as external files to build the text index the literals get assigned continuos TextRecordIDs starting after the last TextRecordID from the external files.
During retrieval the ID Table contains the TextRecordID of a result. If its TextRecordID is bigger than the last TextRecordID of non literals we know it's a literal and thus retrieve differently.
The change was neccessary since counting the nofNonLiterals worked before because the TextRecordIDs in the wordsfiles are continuos but when building with the docsFile they are not. The safer variant is to check for the last TextRecordID directly which is done now.

joka921 replied:
Then add a comment tthat this is only for testing.

Flixtastic committed:
Add method getWordsLineFromDocsFile to simplify parsing in IndexImpl.Text.cpp.

Flixtastic committed:
Reduced complexity of processWordsForInvertedLists.

Flixtastic commented on src/index/Index.h:
```diff
@@ -100,9 +100,10 @@ class Index {
+      std::optional<const string> wordsFile,
+      std::optional<const string> docsFile, bool addWordsFromLiterals = false,
+      bool useDocsFileForVocabulary = false,
+      bool addEntitiesFromWordsfile = false,
```
09036fd1f6bc7b167e47045b24217ef626998542

Flixtastic committed:
Merge branch 'ad-freiburg:master' into docsfile-only-text-index

Flixtastic committed:
Fix error of getWordsLineFromDocsFile

Flixtastic committed:
Fix Sonar issues

Flixtastic committed:
Merge branch 'ad-freiburg:master' into docsfile-only-text-index

Flixtastic committed:
Merge remote-tracking branch 'upstream/master' into docsfile-only-text-index

Flixtastic committed:
Changed naming of fields in TextIndexConfig.h.
Removed faulty error message in IndexBuilderMain.cpp.

joka921 commented on src/index/Index.cpp:
```diff
@@ -26,12 +26,8 @@ void Index::createFromOnDiskIndex(const std::string& onDiskBase,
-  pimpl_->buildTextIndexFile(wordsAndDocsFile, addWordsFromLiterals,
-                             textScoringMetric, bAndKForBM25);
+void Index::buildTextIndexFile(TextIndexConfig config) {
+  pimpl_->buildTextIndexFile(config);
```
I'm sure sonar wants to `std::move` here or use a `const&` whatever you need in the impl.

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -116,26 +99,31 @@ void IndexImpl::logEntityNotFound(const string& word,
-    std::pair<float, float> bAndKForBM25) {
-  AD_CORRECTNESS_CHECK(wordsAndDocsFile.has_value() || addWordsFromLiterals);
+void IndexImpl::buildTextIndexFile(TextIndexConfig config) {
+  config.checkValid();
+  bool addWordsFromFiles =
+      config.docsFile_.has_value() &&
+      (config.wordsFile_.has_value() || config.useDocsFileForVocabulary_);
+  if (config.addOnlyEntitiesFromWordsFile_ &&
+      !config.useDocsFileForVocabulary_) {
+    throw std::runtime_error(
+        "The option to add entities from wordsfile can only be used together "
+        "with the option to build the text index from docsfile since "
+        "otherwise the wordsfile would be used anyway.");
+  }
```
As this code only uses the `config`, it should be in member functions of the config to make the code here easier (especially the part that throws can be part of `checkValid` or something.

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -150,25 +138,30 @@ void IndexImpl::buildTextIndexFile(
   {
-    auto [b, k] = bAndKForBM25;
-    storeTextScoringParamsInConfiguration(textScoringMetric, b, k);
+    auto [b, k] = config.bAndKForBM25_;
+    storeTextScoringParamsInConfiguration(config.textScoringMetric_, b, k);
```
If you are refactoring anyway, you can store the Metric + b + k in a separate struct `textScoreConfig` which is then part of the `TextIndexConfig` which can be passed in here as a single step.

Flixtastic replied:
e64ec2c612fd8f288b437db58a9684361c6d1143

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -150,25 +138,30 @@ void IndexImpl::buildTextIndexFile(
 
   // Build the text vocabulary (first scan over the text records).
-  processWordsForVocabulary(wordsFile, addWordsFromLiterals);
+  processWordsForVocabulary(
+      config.useDocsFileForVocabulary_ ? docsFile : wordsFile,
+      config.addWordsFromLiterals_, config.useDocsFileForVocabulary_);
```
Can you pass in the complete config into this function, you need almost all members
(maybe then the `docsFile` and `wordsFile` variables here become unused.

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -150,25 +138,30 @@ void IndexImpl::buildTextIndexFile(
   TextVec vec{indexFilename + ".text-vec-sorter.tmp",
               memoryLimitIndexBuilding() / 3, allocator_};
-  processWordsForInvertedLists(wordsFile, addWordsFromLiterals, vec);
+  processWordsForInvertedLists(wordsFile, docsFile,
+                               config.addWordsFromLiterals_,
+                               config.useDocsFileForVocabulary_,
+                               config.addOnlyEntitiesFromWordsFile_, vec);
```
Same here, you almsost need the complete config, so pass it in.

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -246,28 +239,43 @@ void IndexImpl::addTextFromOnDiskIndex() {
       distinctWords.insert(line.word_);
     }
+  };
+  if (useDocsFileForVocabulary) {
+    for (const auto& line :
+         getWordsLineFromDocsFile(file, textVocab_.getLocaleManager())) {
+      processLine(line);
+    }
+  } else {
+    for (const auto& line :
+         WordsFileParser{file, textVocab_.getLocaleManager()}) {
+      processLine(line);
+    }
+  }
```
```suggestion
const auto& localeManager = textVocab_.getLocaleManager();
  if (useDocsFileForVocabulary) {
    ql::ranges::for_each(getWordsLineFrom(file, localManager), processLine;
    }
  } else {
    for (const auto& line :
         WordsFileParser{file, textVocab_.getLocaleManager()}) {
      processLine(line);
    }
  }
```
(same refactoring for the otther `for(...){processLine()` loops, saves two lines and increases readability).

Flixtastic replied:
1620441c80164b4545bea8aace49fdf91dbfc8da

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -280,7 +288,10 @@ void IndexImpl::processWordsForInvertedLists(const string& contextFile,
+  auto processLine = [&currentContext, &nofContexts, &vec, &wordsInContext,
+                      &entitiesInContext, &nofEntityPostings, &nofLiterals,
+                      &entityNotFoundErrorMsgCount, &nofWordPostings,
+                      this](const WordsFileLine& line) {
```
see above, probably pass in the complete config to this function.

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -298,26 +309,111 @@ void IndexImpl::processWordsForInvertedLists(const string& contextFile,
+      for (const auto& line :
+           getWordsLineFromDocsFile(docsFile, localeManager)) {
+        processLine(line);
+      }
```
`ql::ranges::for_each` (see above)

Flixtastic replied:
1620441c80164b4545bea8aace49fdf91dbfc8da

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -298,26 +309,111 @@ void IndexImpl::processWordsForInvertedLists(const string& contextFile,
+    // Case where: !useDocsFileForVocabulary
+    for (const auto& line :
+         WordsFileParser{wordsFile, textVocab_.getLocaleManager()}) {
+      processLine(line);
```
see above ( + reuse the localeManager variable).

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -298,26 +309,111 @@ void IndexImpl::processWordsForInvertedLists(const string& contextFile,
+  // Parse literals if specified
+  if (addWordsFromLiterals) {
+    for (const auto& line : wordsInLiterals(currentContext.get() + 1)) {
+      processLine(line);
```
see above.

Flixtastic replied:
1620441c80164b4545bea8aace49fdf91dbfc8da

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -298,26 +309,111 @@ void IndexImpl::processWordsForInvertedLists(const string& contextFile,
+  auto docsFileParser = DocsFileParser{docsFile, localeManager};
+  auto docsFileIterator = docsFileParser.begin();
+  auto wordsFileParser = WordsFileParser{wordsFile, localeManager};
+  auto wordsFileIterator = wordsFileParser.begin();
```
Check that `begin()!= end()` otherwise the dereference is undefined).

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -298,26 +309,111 @@ void IndexImpl::processWordsForInvertedLists(const string& contextFile,
+  // to the respective context in this case the respective document. Once
+  // the wordsfile ID is larger than the docsfile ID the docsfile line is
+  // processed and so on.
+  while (docsFileIterator != docsFileParser.end()) {
+    currentDocsFileLine = *docsFileIterator;
```
Can't this outer loop just be
`for (auto& currentDocsFileLine : docsFileParser)` 
(and then of course delete the begin() calls above)
Can even define the docsFileParser inline in the loop body.

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -298,26 +309,111 @@ void IndexImpl::processWordsForInvertedLists(const string& contextFile,
+  auto docsFileIterator = docsFileParser.begin();
+  auto wordsFileParser = WordsFileParser{wordsFile, localeManager};
+  auto wordsFileIterator = wordsFileParser.begin();
+  // Get the respective lines
+  WordsFileLine currentWordsFileLine = *wordsFileIterator;
+  DocsFileLine currentDocsFileLine = *docsFileIterator;
```
These seem to be dead writes (they are overwritten before they are ever read?)

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -298,26 +309,111 @@ void IndexImpl::processWordsForInvertedLists(const string& contextFile,
+  // processed and so on.
+  while (docsFileIterator != docsFileParser.end()) {
+    currentDocsFileLine = *docsFileIterator;
+    while (wordsFileIterator != wordsFileParser.end()) {
```
Can be a `for` loop without initialization ( `for(; it != end(); ++it)`)
then the loop gets even more readable.

joka921 commented on src/index/IndexImpl.h:
```diff
@@ -556,11 +551,21 @@ class IndexImpl {
-
-  void processWordsForInvertedLists(const string& contextFile,
-                                    bool addWordsFromLiterals, TextVec& vec);
+  size_t processWordsForVocabulary(const string& file,
+                                   bool addWordsFromLiterals,
+                                   bool useDocsFileForVocabulary);
+
+  void processWordsForInvertedLists(const string& wordsFile,
+                                    const string& docsFile,
+                                    bool addWordsFromLiterals,
+                                    bool useDocsFileForVocabulary,
+                                    bool addEntitiesFromWordsFile,
+                                    TextVec& vec);
+
+  template <typename T>
+  void wordsFromDocsFileEntitiesFromWordsFile(
+      const string& wordsFile, const string& docsFile,
+      const LocaleManager& localeManager, T processLine) const;
```
Add some documentation here.

Flixtastic replied:
c912ca3db0ac936f81f2ded4d81a1c4eb9a159b2

joka921 commented on src/index/TextIndexConfig.h:
```diff
@@ -0,0 +1,111 @@
+// Chair of Algorithms and Data Structures.
+// Author: Felix Meisen (fesemeisen@outlook.de)
+
+#ifndef TEXTINDEXCONFIG_H
+#define TEXTINDEXCONFIG_H
```
That is not the correct format for the include guard.

joka921 commented on src/index/TextIndexConfig.h:
```diff
@@ -0,0 +1,111 @@
+  TextScoringMetric textScoringMetric_ = TextScoringMetric::EXPLICIT;
+  std::pair<float, float> bAndKForBM25_ = {0.75f, 1.75f};
+
+  void checkValid() const {
```
You can give this struct an explicit constructor that automatically calls `checkValid()` at the end + 
if it is not needed or valid to change the settings after the fact, then the members should be private + should have explicit getter functions as `const &`.

(It is not a struct, the members are coupled).

joka921 commented on src/parser/WordsAndDocsFileParser.h:
```diff
@@ -192,4 +194,16 @@ class DocsFileParser : public WordsAndDocsFileParser,
   Storage get() override;
 };
 
+inline cppcoro::generator<WordsFileLine> getWordsLineFromDocsFile(
```
document.

Flixtastic replied:
c912ca3db0ac936f81f2ded4d81a1c4eb9a159b2

joka921 commented on src/parser/WordsAndDocsFileParser.h:
```diff
@@ -192,4 +194,16 @@ class DocsFileParser : public WordsAndDocsFileParser,
 
+inline cppcoro::generator<WordsFileLine> getWordsLineFromDocsFile(
+    const string docsFile, const LocaleManager localeManager) {
+  for (const auto& line : DocsFileParser{docsFile, localeManager}) {
+    for (const auto& word :
+         tokenizeAndNormalizeText(line.docContent_, localeManager)) {
+      auto lineToReturn = WordsFileLine{
+          word, false, TextRecordIndex::make(line.docId_.get()), 0, false};
+      co_yield lineToReturn;
+    }
+  }
+}
```
As we want to get rid of coroutines, this can be a combination of 
`ql::ranges::transform`, `ql::ranges::join` etc. , but leave this open to the last, maybe I can help you there if you run into trouble.

Flixtastic replied:
0959e6d813c44f1cecb6039d205f55d4efb2887d

joka921 reviewed (changes requested):
This already looks much cleaner, thank you very much.
Most of my comments are to clean up.

Flixtastic committed:
- Added comment for getLastTextRecordIndexOfNonLiterals
- Added addWordsFromFiles as getter in TextIndexConfig and reduced code in buildTextIndexFile
- Add a TextScoringConfig
- Changed faulty header guards

Flixtastic committed:
- Implement moving the textindexconfig

Flixtastic committed:
- Implement using ql::ranges::for_each() where applicable

Flixtastic committed:
- Implement using ql::ranges::for_each() where applicable
- Refactoring the TextIndexConfig

Flixtastic committed:
- Add documentation

Flixtastic committed:
- Implement using views instead of coroutine in getWordsLineFromDocsFile.

Flixtastic committed:
- Fix Sonar issues

Flixtastic committed:
Merge branch 'ad-freiburg:master' into docsfile-only-text-index

Flixtastic committed:
- Fix error

Flixtastic committed:
Merge branch 'docsfile-only-text-index' of github.com:Flixtastic/qlever into docsfile-only-text-index

Flixtastic committed:
- Fix error

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -99,31 +99,21 @@ void IndexImpl::logEntityNotFound(const string& word,
-  string wordsFile = config.wordsFile_.value_or("");
-  string docsFile = config.docsFile_.value_or("");
+void IndexImpl::buildTextIndexFile(TextIndexConfig&& textIndexConfig) {
+  const auto config = TextIndexConfig(std::move(textIndexConfig));
```
That is not really necessary.
The parameter can just be `TextIndexConfig config` (no `&&` , but still consistently `std::move()`).

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -137,31 +127,23 @@ void IndexImpl::buildTextIndexFile(TextIndexConfig config) {
-    auto [b, k] = config.bAndKForBM25_;
-    storeTextScoringParamsInConfiguration(config.textScoringMetric_, b, k);
-  }
+  { storeTextScoringParamsInConfiguration(config.getTextScoringConfig()); }
```
```suggestion
  storeTextScoringParamsInConfiguration(config.getTextScoringConfig());
```

joka921 commented on src/index/IndexImpl.Text.cpp:
```diff
@@ -371,13 +344,12 @@ void IndexImpl::wordsFromDocsFileEntitiesFromWordsFile(
-  // Get the respective lines
-  WordsFileLine currentWordsFileLine = *wordsFileIterator;
-  DocsFileLine currentDocsFileLine = *docsFileIterator;
+  AD_CORRECTNESS_CHECK(wordsFileParser.begin() != wordsFileParser.end(),
+                       "When adding entities from wordsfile the given "
+                       "wordsfile can't be an empty file.");
```
```suggestion
  AD_CORRECTNESS_CHECK(wordsFileIterator.begin() != wordsFileParser.end(),
                       "When adding entities from wordsfile the given "
                       "wordsfile can't be an empty file.");
```
(if WordsFilePraser is an inputRange like a generator, then calling begin() multiple times can be a problem).

joka921 commented on src/index/IndexImpl.h:
```diff
@@ -551,17 +551,22 @@ class IndexImpl {
+  // occurrences)
+  size_t processWordsForVocabulary(const TextIndexConfig& textIndexConfig);
+
+  // Fills the given vector with all occurrences of entities and words in the
```
```suggestion
  // Fills the given `vec` with all occurrences of entities and words in the
```

joka921 commented on src/index/TextIndexConfig.h:
```diff
@@ -64,20 +63,50 @@
  *                                         0 <= b <= 1 and 0 <= k.
  */
-struct TextIndexConfig {
+class TextIndexConfig {
```
Some documentation, especially for the getters that contain logic.

joka921 commented on src/parser/WordsAndDocsFileParser.h:
```diff
@@ -194,16 +194,20 @@ class DocsFileParser : public WordsAndDocsFileParser,
+                                      0, false};
+               });
+      });
+  return wordLines | ql::views::join;
```
That is definitely dangling, because `wordLines` is captured by reference into the `join`, 
but you said that there is a problem here, I'll have a look at it.

joka921 reviewed (changes requested):
Some smaller comments, but overall this looks nice.

joka921 committed:
Fix the compilation.

Signed-off-by: Johannes Kalmbach <johannes.kalmbach@gmail.com>

joka921 committed:
Merge in the master.

Signed-off-by: Johannes Kalmbach <johannes.kalmbach@gmail.com>

Flixtastic committed:
Implement small requested changes

Flixtastic committed:
Refactor the code for building the text index (#1949)

Refactor of the code for building (and using) the text index, for improved readability and maintainability. No change in functionality. In particular, creates two new files `TextIndexBuilder.h` and `TextIndexBuilder.cpp`.

Flixtastic committed:
Merge branch 'temp-merge' into docsfile-only-text-index

Flixtastic committed:
Merge remote-tracking branch 'upstream/master' into docsfile-only-text-index

# Conflicts:
#	src/index/IndexBuilderMain.cpp
#	src/index/TextIndexBuilder.cpp
#	src/index/TextIndexBuilder.h
#	test/GroupByTest.cpp
#	test/util/IndexTestHelpers.cpp

Flixtastic committed:
Fix Sonar Issues

Flixtastic committed:
Fix Sonar Issues and Clang 18 compiling problemsgit

Flixtastic committed:
Fix missed error of Clang 18 compiling

Flixtastic committed:
Fix Sonar Issues and Code Cov.

Flixtastic committed:
Remove unused parameters of lambda.

Flixtastic committed:
Fix Sonar Issues

Flixtastic committed:
Merge branch 'master' into docsfile-only-text-index

Flixtastic committed:
Merge branch 'ad-freiburg:master' into docsfile-only-text-index

joka921 committed:
Merge branch 'master' into docsfile-only-text-index

# Conflicts:
#	src/index/TextIndexBuilder.cpp

joka921 reviewed (approved):
Thank you very much!

Flixtastic committed:
Merge ad-freiburg/qlever into branch.

hannahbast reviewed (changes requested):
@Flixtastic Thank you for this PR. I tried it, using the IMDb dataset, and encountered an error. First, I produced a file `imdb.docsfile` (based on a SPARQL query, but that shouldn't be important) with 11,536,090 lines like the following:
```
1       Carmencita
2       Le clown et ses chiens
3       Poor Pierrot
4       Un bon bock
5       Blacksmith Scene
```
Then I build an index using the following command-line:
```
echo '{ "num-triples-per-batch": 1000000 }' > imdb.settings.json
cat imdb.ttl | IndexBuilderMain -i imdb -s imdb.settings.json --vocabulary-type on-disk-compressed -F ttl -f - -p true -d imdb.docsfile | tee imdb.index-log.txt
```
Then I started the server as follows:
```
ServerMain -i imdb -j 8 -p 7029 -m 5G -c 2G -e 1G -k 200 -s 300s
```
Then I asked the following query:
```
curl -s drau:7029 -H "Accept: text/tab-separated-values" --data-urlencode query='SELECT ?text WHERE { ?text ql:contains-word "matrix" } LIMIT 10'
```
This yielded the following exception:
```
Assertion `subset.empty() || subset.back() < numColumns()` failed. Please report this to the developers. In file \"/local/data-ssd/qlever/qlever-code/src/engine/idTable/IdTable.h \" at line 626
```

Flixtastic committed:
Add more information during text index building. Make the -D and -E options during index building bool switches instead of value boost options.

Flixtastic committed:
Merge upstream/master into branch. Use the new test struct in TextIndexScanForWordTest for the docsfile only test.

Flixtastic committed:
Merge branch 'ad-freiburg:master' into docsfile-only-text-index

Flixtastic committed:
Merge remote-tracking branch 'upstream/master' into docsfile-only-text-index

joka921 reviewed (changes requested):
if (as @hannahbast tried) only a docsfile is specified, and neither the `-D` option (read the words from the docsfile) or an wordsfile is given, this should immediately (in the index builder) raise an exception that tells us very verbosely exactly this.

Flixtastic committed:
Add check that makes sure the flag to build from docsfile is set if only the docsfile was sepcified.

Flixtastic committed:
Move the throw to an earlier position in code before building the normal index.

joka921 reviewed (approved):
Thank you very much.

Flixtastic committed:
Merge branch 'ad-freiburg:master' into docsfile-only-text-index

hannahbast reviewed (approved):
Approving this, so it appears in the list of approved PRs

Flixtastic committed:
Merge upstream/master into branch.

hannahbast reviewed (commented):
@Flixtastic and Cc @joka921  I had another look at this PR, here are some comments:

1. Please write a proper description that explains how to use the new feature.

2. In particular, the semantics of the new `-D` option is not properly explained. I first tried to use it as follows: `IndexBuilderMain ... -D test.docsfile.tsv`. The result was that no text index was built but also no error message was shown. I then realized that I should call `IndexBuilderMain ... -d test.docsfile.tsv -D`

3. I build a test index using `scientists.nt` from our E2E tests and a `test.docsfile.tsv` with 1M text records obtained via a SPARQL query on Wikidata (English descriptions of entities). I then tried this query: https://qlever.cs.uni-freiburg.de/scientists/nsrnIK . The number of results seems to be correct ( I verified with an equivalent `grep` on `test.docsfile.tsv`), but the displayed texts don't make sense

A general remark that is also relevant for your other PR: Please make sure that the description and comments can be understood by someone who is reasonably familiar with QLever, but not with this particular PR

Flixtastic committed:
Provide better description for the flags of the docsfile only building. Catch faulty options given.

Flixtastic committed:
Merge upstream/master into branch.

Flixtastic committed:
Fix merge errors

joka921 reviewed (changes requested):
Some small suggestions for cleanups.

Have you seen the bug report by @hannahbast for a text index not working as expected with the texts etc?

Flixtastic committed:
Change wordings of error messages. Add tests for docsfile only building config validate.
